cornac/models/dmrl/pwlearning_sampler.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2018 The Cornac Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+
+from typing import List
+import torch.utils.data as data
+from cornac.data.dataset import Dataset
+import numpy as np
+
+
+class PWLearningSampler(data.Dataset):
+    """
+    Sampler for pairwised based ranking loss function. This sampler will return
+    a batch of positive and negative items. Per user index there will be one
+    positive and num_neg negative items.
+    This sampler is to be used in the PyTorch DatLoader, through which loading
+    can be distributed amongst multiple workers.
+    """
+    def __init__(self, cornac_dataset: Dataset, num_neg: int):
+        self.data = cornac_dataset
+        self.num_neg = num_neg
+        # make sure we only have positive ratings no unseen interactions
+        assert np.all(self.data.uir_tuple[2] > 0)
+        self.user_array = self.data.uir_tuple[0]
+
+        self.item_array = self.data.uir_tuple[1]
+        self.unique_items = np.unique(self.item_array)
+        self.unique_users = np.unique(self.user_array)
+        self.user_item_array = np.vstack([self.user_array, self.item_array]).T
+        # make sure users are assending from 0
+        np.all(np.unique(self.user_array) == np.arange(self.unique_users.shape[0]))
+
+    def __getitems__(self, list_of_indexs: List[int]):
+        """
+        Vectorized version of __getitem__
+
+        Uses list_of_indixes to index into uir_tuple from cornac dataset and
+        thus retrieve 1 positive item per given user. Additionally random
+        samples num_neg negative items per user given in list_of_indices.
+
+        Parameters
+        ----------
+
+        list_of_indexs: List[int]
+            List of indexs to sample from the uir_tuple given in cornac dataset
+        """
+        batch_size = len(list_of_indexs)
+        users = self.user_array[list_of_indexs]
+        pos_items = self.item_array[list_of_indexs]
+
+        pos_u_i = np.vstack([users, pos_items]).T
+
+        # sample negative items per user
```

Comment:
This is similar to the [Dataset.uij_iter()](https://github.com/PreferredAI/cornac/blob/c6035980a3ee0735261f8327b9c3a1da870f7de8/cornac/data/dataset.py#L490) which I believe it could be reused for simpler implementation.

Reply:
Not exactly what's needed here as uij_iter() returns a set of users, 1 positive and 1 negative array. For DMRL we need a set of 1 user array, 1 pos item array and num_neg (e.g. 4) negative items per user in a 2D array.
But thanks for the hint! I did take a look at it and it is indeed not far off.

Reply:
Got it. I overlooked the number of negative samples